akka-persistence/src/main/scala/akka/persistence/journal/leveldb/LeveldbStore.scala

```diff
@@ -42,9 +42,24 @@ private[persistence] trait LeveldbStore extends Actor with WriteJournalBase with
 
   def asyncWriteMessages(messages: immutable.Seq[AtomicWrite]): Future[immutable.Seq[Try[Unit]]] =
     Future.fromTry(Try {
-      withBatch(batch ⇒ messages.map { a ⇒
-        Try(a.payload.foreach(message ⇒ addToMessageBatch(message, batch)))
-      })
+      withBatch { batch ⇒
+        var result: immutable.Seq[Try[Unit]] = Nil
+        val iter = messages.iterator
+        var n = 0
+        while (iter.hasNext) {
+          try {
+            iter.next().payload.foreach(message ⇒ addToMessageBatch(message, batch))
+            n += 1
+          } catch {
+            case NonFatal(e) ⇒
+              if (result.isEmpty)
+                result = Vector.fill(n)(AsyncWriteJournal.successUnit)
+              result :+= Failure(e)
+          }
+        }
+
+        result
+      }
```

Comment:
The purpose with the PR is to support Future(Nil) in asyncWriteMessages so that implementors of journal plugins can use that for optimizations (reduce allocations). It was requested by some (and we also discussed it early).